Vertical arrow key navigation

ellatrix opened the pull request:
## Description
This is a rebase of #2296.

It allows the caret in a block to be vertically repositioned across focusable elements.

## How Has This Been Tested?
In e.g. the demo content, place the caret in an editable container. Use the UP and DOWN arrow keys to navigate across editable areas and blocks.

## Checklist:
- [x] My code is tested. (No unit test because it uses layout.)
- [x] My code follows the WordPress code style.
- [x] My code follows has proper inline documentation.

iseulde commented:
There's an issue with positioning the caret when you try to move it up vertically and the toolbar is in the way... Then it will try. to focus a button in the toolbar. 😄

iseulde commented:
So far it works perfectly in Chrome and Safari. Firefox is a complete mess again.

iseulde commented:
Getting quite happy with this. Squashed down a few other bugs. On rare occasions the browser fails to move position, have yet to figure out these few edge cases.

iseulde commented:
Fixed an issue where the new range is not within the target container.
Fixed an issue where vertical arrow keys wouldn't work in an empty container.

iseulde commented:
Encountered one more bug with the button block. It will attempt to put the caret next to it but fail. :) Will fix in a bit.

jasmussen commented:
This seems to be working QUITE well. Impressive. Lovely. 

As you get near the end of the demo content, the page seems to scroll in jumps, so to speak. This can make it a little easy to lose focus of where the caret is. This is very possibly just normal browser behavior, in which case _ignore crazy old Joen_. But otherwise — what could this be and is there a way to make it smoother, perhaps that typewriter approach?

jasmussen commented:
Inside the code block, the range stuff doesn't work. Which is totally fine, it doesn't have to — but figured I'd mention it in case it's like a one line code change.

iseulde commented:
I think that might be the bug I mentioned. I know you'd like that typewriter feel in general, but let's attempt that in a separate PR. 😄
U'll look at the code block. We have to put at the start or end there, there's no way around that for inputs and text areas. :/

jasmussen commented:
> I know you'd like that typewriter feel in general, but let's attempt that in a separate PR. 😄

Totally, no worries.

> I'll look at the code block. We have to put at the start or end there, there's no way around that for inputs and text areas. :/

No worries — this enhancement is way more useful for the editables. Doesn't have to work elsewhere imo.

mcsf commented on editor/utils/dom.js:
```diff
@@ -56,18 +57,85 @@ export function isEdge( container, start = false ) {
+	const rangeRect =
+		range.startContainer.nodeType === window.Node.ELEMENT_NODE ?
+		range.startContainer.getBoundingClientRect() :
+		range.getClientRects()[ 0 ];
```
Minor point, but I find that multi-line ternaries read better like:

```js
a
  ? b
  : c
```

but currently our linter has no preference.

iseulde replied:
Fair enough. I've been used to this way. Maybe we should add a rule so that it's the same everywhere?

EphoxJames replied:
My only worry is that the Javascript interpreter of some browser out there will see the condition as the end of the statement if the question mark is not the same line. It may be changed in transpiling though so if Babble gets it right then I suppose I don't need to worry.

iseulde replied:
That's probably why I've been used to it this way.

mcsf replied:
I had a whole reply with an excerpt of the [ECMA spec](https://www.ecma-international.org/ecma-262/8.0/index.html#sec-automatic-semicolon-insertion) and an analysis making the case that only a faulty browser would be too eager to insert a semicolon in this case, but then I found the right Crockford [source](http://javascript.crockford.com/code.html). :)

> The ternary operator can be visually confusing, so ? question mark always begins a line and increase the indentation by 4 spaces, and : colon always begins a line, aligned with the ? question mark. The condition should be wrapped in parens.

My take is that Crockford has historically been a conservative reference on JS (and, [apparently](https://stackoverflow.com/a/7259181), a previous version of the same text had advised against having an operator at the beginning of a line), and thus his OK means that we're totally safe here and that it's ultimately a stylistic decision. How do y'all weigh in?

iseulde replied:
I don't have a strong opinion. If browsers are a worry, we anyway process with Babel and everything gets minified.

mcsf commented on editor/utils/dom.js:
```diff
@@ -77,17 +145,101 @@ export function placeCaretAtEdge( container, start = false ) {
+	// Editable was already focussed, it goes back to old range...
+	// This fixes it.
+	selection.removeAllRanges();
+	selection.addRange( range );
```
Could you provide more detail on the need to repeat this? Is it a quirk?, is it something otherwise preventable?, etc.

iseulde replied:
I suspect it might be related to TinyMCE's `editor.selection.lastFocusBookmark`, but I need to investigate. Maybe @ephox-mogran or @EphoxJames know more.

EphoxJames replied:
I don't know off the top of my head but I will dig into the code and see if I can find out.

iseulde replied:
Thanks! Ideally we wouldn't have to touch the TinyMCE editor instance at all here.

mcsf commented on editor/utils/dom.js:
```diff
@@ -77,17 +145,101 @@ export function placeCaretAtEdge( container, start = false ) {
+
+	const range = caretRangeFromPoint( document, x, y );
+
+	container.style.zIndex = null;
```
Hm, bummer. :)

Local testing (on this very page) indicates that `container.style.pointerEvents = 'none';` is just as effective. Should we prefer that one instead? I wonder if that way we minimize our chances of generating visual artefacts.

iseulde replied:
How would that be used? On the toolbar element? That would require us to have the toolbar element, which would be less isolated logic. In that case I'd prefer the z-index approach.

mcsf replied:
Yes, it would be on the toolbar — I was looking at this so closely that I didn't weigh the drawback of having to deal with yet another entity. z-index may be the lesser evil, then.

mcsf commented on editor/utils/dom.js:
```diff
@@ -56,18 +57,85 @@ export function isEdge( container, start = false ) {
+ *
+ * @type {DOMRect}
+ */
+let firstVerticalRect;
```
My biggest concern is this global here. It's a concern in general, but aggravated by the fact that we are in a `utils` module, where the expectation of lingering state is low, _and_ that the main function that sets it is a seemingly innocent predicate, `isVerticalEdge`.

I see it as a definite code smell that we should tackle, probably before merging. Does it mean that the predicate should become something else? That it should be broken into different pieces? That it should become a function returning a `DOMRect` or `null`? Relatedly, `onMouseDown={ resetVerticalPosition }` on WritingFlow is an implementation detail that shouldn't be leaking.

iseulde replied:
Any ideas would we welcome! I'll have a think.

ephox-mogran replied:
Can firstVerticalRect just be passed in, and have its state / value managed in WritingFlow itself? It is essentially part of the state of writing flow, if that's where we are doing the cursor movement.

iseulde replied:
Okay, I changed this.

mcsf commented on editor/utils/dom.js:
```diff
@@ -56,18 +57,85 @@ export function isEdge( container, start = false ) {
  */
-export function placeCaretAtEdge( container, start = false ) {
+export function isVerticalEdge( container, reverse ) {
+	if ( [ 'INPUT', 'TEXTAREA' ].indexOf( container.tagName ) !== -1 ) {
```
Minor: naming this, as is done in other parts of the diff, would be nice: `const isInputOrTextarea =`

iseulde replied:
I almost changed the other one. :) What makes you prefer it?

mcsf replied:
That `indexOf` isn't very descriptive and the variable assignment counters that. I would also go, instead, for `lodash#includes`.

mcsf commented on editor/writing-flow/index.js:
```diff
@@ -37,31 +43,63 @@ class WritingFlow extends Component {
+		const right = keyCode === RIGHT;
+		const reverse = up || left;
+		const horizontal = left || right;
+		const vertical = up || down;
```
As some of the functions here and in `dom/utils` grow, keeping track of boolean variables becomes harder. Aside from a `[modal]+[main verb]` nomenclature (with _modal_ taking values in _{  is, should, can, may, will, … }_),  I'd love to see more descriptive names throughout. Examples:

- `up` → `isUpKey`
- `reverse` → `isAscending`
- `noScroll` → logic reversal, becoming `mayUseScroll`

mcsf reviewed (changes requested):
Here's a first review. I think we should tackle the matter of `firstVerticalRect`.

I will now switch context for a bit so that I can come back to this later with a clearer head and see if I can help.

iseulde commented on editor/writing-flow/index.js:
```diff
@@ -37,31 +43,63 @@ class WritingFlow extends Component {
 		const focusableNodes = this.getVisibleTabbables();
-		if ( direction === 'UP' ) {
+
+		if ( reverse ) {
```
I don't really see the problem here? Should I change it?

ephox-mogran commented:
@iseulde I've found one issue that might be making it a little inconsistent. It's most visible when using the table block. If you setup a paragraph below the table and after the table, and try to navigate from below the table up, or above the table down, then the caret doesn't stay in the same horizontal position when it is entering the table.

This is because of two reasons:

1. There is padding on the table cells which interferes with y positions of nodes. This directly impacts (2)
2. The `caretRangeFromPosition` method can return a range that has a bounding rectangle that **does not include** the (x, y) coordinate passed in. It is a 'close-enough' type of API that is designed for finding the nearest text. The problem is that when it finds a text node, but the (x, y) is not actually inside that text node, it gets the offset wrong and just chooses either the start of the node or the end of the node, which is not what you want.

In order to offset this, you can run the caretRangeFromPosition method again, but this time, use an (x, y) that is bounded within the `getBoundingClientRect` of the range it returned initially. In this case, just use the same `x` and use a `y` that is either the rect.top or rect.bottom (or potentially offset that number by one or two). This then means that it is looking within the text itself (and not inside the padding), and it will give you an accurate offset.

Now, you may get a problem where it will just keep recursing indefinitely, because it can't find a text position that is any better than it's first guess. Therefore, you should quit after some number of retries. It shouldn't be very many.

Note, the cursor definitely doesn't keep the same position when moving through the table itself, but that should be logged against TinyMCE.
```

iseulde commented:
I'll have a look at the table block.

iseulde commented:
For me going from a paragraph up to a table seems to work fine. Going from a paragraph down to a table puts it in the right cell, but not the right horizontal position. The padding probably plays a role in this, as you said. Ideally this should have margin instead, not padding?

iseulde commented:
Sorry, ignore me about the margin, I'm confusing it with another case...

iseulde commented:
@ephox-mogran I pushed a fix. Is this similar to what you had in mind?

ellatrix committed:
Add vertical arrow navigation for blocks

iseulde commented:
Okay, as far as I can see, there are two issues remaining:

* Sometimes ranges are not set (no idea what the reason is), or they are set at the last focussed range. The latter is fixed with a hack for vertical navigation, so either we add this hack for horizontal too, or we find a better way to fix this.
* In the master branch, as well as in this branch, reverse horizontal navigation breaks if there is formatting. Press the left arrow key at the start of a paragraph, and it will end of at the end of a formatting tag in the previous editable. As far as I see, this seems to be caused by `range.collapse()`.

mcsf committed:
Refactor to slim down WritingHelper#keyDown

- Split `getVerticalEdge` into two separate helpers: `computeCaretRect`
  and `isVerticalEdge`.

- Rearrange the logic in `WritingFlow#keyDown` to make the symmetries
  between horizontal- and vertical-navigation handlings more apparent,
  as well as the lifecycle of `this.verticalRect`.

iseulde commented on editor/utils/dom.js:
```diff
@@ -101,32 +97,47 @@ export function getVerticalEdge( container, isReverse ) {
-		isEdge: true,
-	};
+	// Adjust for empty containers.
+	const caretRect =
```
Any reason to not return right away? :)

mcsf replied:
Mostly vestigial :) but it seemed to make sense to pair the pre-existing comment with the const assignment. I can push a direct `return`.

mcsf committed:
placeCaretAtVerticalEdge: fix TypeError when target out of view

mcsf commented:
@iseulde: a224863c fixes an occasional bug I was seeing, which can be reproduced consistently this way:
1. Open the Gutenberg Demo
2. Place the cursor at `Of mountains & →|←printing presses`
3. Keep hitting the down arrow
4. An error should be thrown when switching into block "... like this one, which is right aligned."

ephox-mogran commented:
I'm looking into seeing why 

> In the master branch, as well as in this branch, reverse horizontal navigation breaks if there is formatting. Press the left arrow key at the start of a paragraph, and it will end of at the end of a formatting tag in the previous editable. As far as I see, this seems to be caused by range.collapse().

is happening.

ephox-mogran commented:
@iseulde , it looks like what you are seeing is a bug in tinymce. I've recreated it in a fiddle here (http://fiddle.tinymce.com/Scgaab/2). As soon as you remove the `tinymce.init` call, it works again.

mcsf commented:
@iseulde & @ephox-mogran, drive-by comment before I have to run: breaking on `selectionchange` yields the following:

<img width="788" alt="screen shot 2017-10-23 at 18 35 48" src="https://user-images.githubusercontent.com/150562/31903779-23ca2f02-b821-11e7-95ee-a5b1fbe392d5.png">

mcsf commented on editor/writing-flow/index.js:
```diff
@@ -37,31 +45,55 @@ class WritingFlow extends Component {
 	}
 
 	onKeyDown( event ) {
 		const { keyCode, target } = event;
```
I missed this until now, but FYI [`KeyboardEvent.keyCode` has been deprecated](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/keyCode) and `KeyboardEvent.key` is to be preferred.

Something to revise in the future?

iseulde replied:
Probably. We've been using it in Gutenberg quite a lot, so does TinyMCE. Maybe better to discuss separately.

aduth replied:
Previously I'd avoided `key` because of browser support; I don't recall if this is specifically an older IE thing, or due to the non-standard behavior described in the "Notes" at:

http://caniuse.com/#feat=keyboardevent-key

iseulde commented:
@ephox-mogran Yeah, the fiddle illustrates this well. Only breaks in Chrome for me, not Safari.

mcsf commented:
Re

> In the master branch, as well as in this branch, reverse horizontal navigation breaks if there is formatting. Press the left arrow key at the start of a paragraph, and it will end of at the end of a formatting tag in the previous editable. As far as I see, this seems to be caused by range.collapse().

Culprit appears to be a fix for browser quirks:

https://github.com/tinymce/tinymce/blob/d2c9a9a03e3e2adabcd01a46d65640e49f5beb75/src/core/src/main/js/util/Quirks.js#L529-L544

> Fixes selection issues where the caret can be placed between two inline elements like `<b>a</b>|<b>b</b>` this fix will lean the caret right into the closest inline element.

I didn't spot any obvious override mechanism to get around this quirk fix. I played a bit with event interception and `tinymce.init({ setup })`, to no avail. @ephox-mogran, short of pushing an upstream change, is there any sort of TinyMCE-fu that could help us here? :)

ephox-mogran commented:
@mcsf the fix in normalisation of ranges for this is in functional review in the TinyMCE development team. Outside of waiting for that fix for core to be merged, there's not much you can do. It's a high priority.

ephox-mogran commented:
@mcsf , @iseulde , how many blocking issues are there still with this PR? Are there any we have replication cases for (aside from the normalisation range bug in TinyMCE above?)

Specifically, is this still happening?

> Sometimes ranges are not set (no idea what the reason is), or they are set at the last focussed range. The latter is fixed with a hack for vertical navigation, so either we add this hack for horizontal too, or we find a better way to fix this.

mcsf commented:
> the fix in normalisation of ranges for this is in code review in the TinyMCE development team. 

Great to know! Can you keep us posted? Seems like we can hold off this PR till EOW, perhaps?

> Specifically, is this still happening?

I don’t think I ever ran into that one. 

> how many blocking issues are there still with this PR?

None on my part. How about you, @iseulde?

ephox-mogran commented:
> Great to know! Can you keep us posted? Seems like we can hold off this PR till EOW, perhaps?

No problem.

Also, given that this bug is already in master, and has to do with horizontal navigation and not vertical navigation, it probably shouldn't hold up this PR which is focusing on vertical navigation. Especially, as it's an internal TinyMCE bug. Would you agree @mcsf ?

mcsf commented:
Good point. We should also rebase and retest.

ephox-mogran reviewed (approved):
Aside from my issue with using textContent for non text nodes when dealing with offsets (which is already in master, so not related to this PR), it looks good to me.

ephox-mogran commented:
@mcsf should I rebase #3038 now based on this and update the PR? Or do you still have another approach in mind that you want to try first?

mcsf commented:
You can go ahead and rebase, @ephox-mogran. Thanks!

aduth commented on editor/writing-flow/index.js:
```diff
@@ -71,7 +103,9 @@ class WritingFlow extends Component {
 				ref={ this.bindContainer }
-				onKeyDown={ this.onKeyDown }>
+				onKeyDown={ this.onKeyDown }
+				onMouseDown={ () => this.verticalRect = null }
```
It would have been good to assign this to a bound instance method so that it's not creating a new function reference (and thereby applying a change diff) on every render.

mcsf replied:
Darn, I'm usually stringent about those.

ephox-mogran commented:
@iseulde @mcsf 

> In the master branch, as well as in this branch, reverse horizontal navigation breaks if there is formatting. Press the left arrow key at the start of a paragraph, and it will end of at the end of a formatting tag in the previous editable. As far as I see, this seems to be caused by range.collapse().

This has been fixed as of TinyMCE 4.7.2.

aduth commented on editor/writing-flow/index.js:
```diff
@@ -20,6 +26,8 @@ class WritingFlow extends Component {
 		this.onKeyDown = this.onKeyDown.bind( this );
 		this.bindContainer = this.bindContainer.bind( this );
+
+		this.verticalRect = null;
```
Arriving at this pull request hopefully confused about what this variable is. Can we add some inline comments?

mtias replied:
Might be good to make an issue :)

iseulde replied:
Comment coming.
